content/stdout-vs-stderr.md

```diff
@@ -0,0 +1,2077 @@
+
+In the screencast below, I modified the `crossterm` backend of `ratatui` to highlight the cells that are not being changed between renders. You can clearly see from the number of red cells that we are not able skip a lot of `write` calls since mostly everything is changing on the terminal:
+
+![stdout highlighted](stdout-highlighted.gif)
+
+However this is not the case for most of the TUI applications and this optimization actually saves us from re-rendering the majority of the screen.
+
+Another interesting thing to look at is the buffer size. We can observe the following with a smaller buffer size (100 bytes) and delay between renders:
+
+![stdout small buffer](stdout-small-buffer.gif)
+
+Whereas a bigger buffer renders bigger chunks:
+
+![stdout big buffer](stdout-big-buffer.gif)
+
+We won't see a big difference if we remove the sleep except if we use very small/big buffer then the FPS drops vastly. We can probably experiment with the buffer size to draw a single line for each render but I wasn't able to get a better FPS in my attempts.
+
+One other thing to mention, there were [recent developments](https://github.com/ratatui-org/ratatui/pull/601) on `ratatui` to achieve a better performance for rendering cells. However, this doesn't have a big impact on FPS but definitely an improvement for using less resources.
+
+We can keep experimenting with the low level functions of `crossterm` / `ratatui` to further optimize things but I feel like that would be a better topic for the **part 2** of this post.
+
+While writing this, I wasn't able to achieve a "faster stdout" so feel free to leave comments about your suggestions!
+
+---
+
+## <g>Findings</g>
+
+Here is the `ratatui` + `crossterm` rendering comparison for stdout and stderr using unbuffered / line-buffered / block-buffered writes:
+
+![stdout vs stderr detailed](stdout-vs-stderr-detailed.gif)
+
+<center>
+<span class="glowy-code">
+
+[stdout-vs-stderr-all.rs](https://github.com/orhun/rust-stdout-vs-stderr/blob/main/src/stdout-vs-stderr-all.rs)
+
+</span>
+</center>
+
+The takeaway from this is that I/O streams have similar performances when the same buffering technique is used. We can also say that `std::io::stdout()` is <g>faster</g> than `std::io::stderr()` because the use of line-buffered vs no buffering.
+
+If you want to reproduce the same results, I used the following environment in my experiments:
+
+- Backend: [crossterm](https://github.com/crossterm-rs/crossterm)
+- TUI/rendering: [ratatui](https://github.com/ratatui-org/ratatui)
+- Terminal: [alacritty](https://github.com/alacritty/alacritty)
+- System: 8 core CPU + 24gb RAM
+
+I'm curious about how the results might change in different systems/terminals so feel free to share your findings below!
```

Comment:
Since Alacritty is GPU accelerated, your GPU might be interesting?

Reply:
added!